feat(gnovm): add `Coin` constructor and more functionality

leohhhn opened the pull request:
<!-- please provide a detailed description of the changes made in this pull request. -->

## Description

Superseding https://github.com/gnolang/gno/pull/1696




<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

leohhhn commented:
I have some thoughts about this:

Coins and their denoms are validated in the TM2 layer. The question is - should we also implement this validation in Gno, for example:
```
func NewCoin(denom string, amount int64) Coin {
    if err := validate(denom, amount); err != nil {
       panic(err)
    }

    return Coin{
       Denom:  denom,
       Amount: amount,
    }
}
```
or, is it maybe a better option to not do "double" validation since it spends gas? basically leave the validation and panicking to TM2 instead of the VM?

Need some input.

leohhhn commented:
While we're on this topic, I came across this issue I made some time ago: https://github.com/gnolang/gno/issues/1676

Thinking about this a bit, we should really make a good, smooth constructor for the `Coins` struct.

One approach that I have is the following:

```
func NewCoins(val ...interface{}) Coins {
	if len(val)%2 != 0 {
		panic("invalid number of arguments")
	}

	cz := Coins{}
	for i := 0; i < len(val)-1; i++ {
		denom := val[i].(string) // todo validate?
		amt := val[i+1].(int64) // todo validate?
		cz = append(cz, NewCoin(denom, amt))
	}

	return cz
}
```

I'm aware that this is not the safest approach, but its probably the most comfortable to use. With this, you can just do 

```
coins := std.NewCoins("ugnot", 100, "othercoin", 200)
```

and its very painless, instead of `coins:= std.Coins{std.Coin{ "ugnot", coinToSend }})`.

WDYT?

ltzmaxwell commented:
> I have some thoughts about this:
> 
> Coins and their denoms are validated in the TM2 layer. The question is - should we also implement this validation in Gno, for example:
> 
> ```
> func NewCoin(denom string, amount int64) Coin {
>     if err := validate(denom, amount); err != nil {
>        panic(err)
>     }
> 
>     return Coin{
>        Denom:  denom,
>        Amount: amount,
>     }
> }
> ```
> 
> or, is it maybe a better option to not do "double" validation since it spends gas? basically leave the validation and panicking to TM2 instead of the VM?
> 
> So, a constructor in Gno would be:
> 
> ```
> func NewCoin(denom string, amount int64) Coin {
>     return Coin{
>        Denom:  denom,
>        Amount: amount,
>     }
> }
> ```
> 
> Need some input.


The validation process in coin.go occurs during both instantiation and operations like Add and Subtract. If validation isn't conducted within constructor of coin.gno, it'll be deferred to tm2 coin.go, leading to additional gas consumption due to extra operations. However, the efficiency depends on how frequently invalid denominations or amounts are set. This consideration is pivotal in determining the optimal approach. These are just initial thoughts, so I'm not entirely certain.

leohhhn commented:
@ltzmaxwell thank you for your answer - it's possible that I wasn't fully clear - currently, there is no validation in the `coins.gno` file as it doesn't exist. So all validation is done in TM2 currently. The question is the following:

If we choose to add validation in `coins.gno`, will that not introduce double validation (on both the TM2 and Gno level) and thus use up more (double) the gas?

My question is the following: should we just not check at the Gno level, and leave the errors & panics to TM2 in this case, exactly to avoid using more gas than needed?

I will play around with this and measure the gas usage in both cases.

BTW, do you have any opinions on [this](https://github.com/gnolang/gno/pull/2104#issuecomment-2113029506) comment?

ltzmaxwell commented:
> If we choose to add validation in `coins.gno`, will that not introduce double validation (on both the TM2 and Gno level) and thus use up more (double) the gas?

> My question is the following: should we just not check at the Gno level, and leave the errors & panics to TM2 in this case, exactly to avoid using more gas than needed?
> 
> I will play around with this and measure the gas usage in both cases.
> 

what I see in tm2/pkg/std.coin.go, the validation happens in `Constructor` and `IsValid()`.

assume we are going to initiate a new coin with denominations of "Ugnot", which is obviously invalid,

- If validation is implemented at the gno level, the system will immediately identify the error and halt execution;
-  Alternatively, if there is no gno level validation, the error will be caught later during the IsValid() check, leading to a halt in execution, albeit through a longer process.

On the other hand, I believe our priorities should be arranged as follows: security takes precedence, followed by readability and maintainability, and finally, gas consumption.

ajnavarro commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,21 +15,77 @@ type Coin struct {
+
+// String provides a human-readable representation of a coin
 func (c Coin) String() string {
+	if c.IsZero() {
+		return ""
+	}
```
shouldn't the output be `0DENOM`? I see misleading outputs if we use this method to show info to the user (example: `your current bank account status:`  instead of `your current bank account status: 0DENOM`)

leohhhn replied:
Here the issue is the following

if you use

leohhhn replied:
Fair point

ajnavarro commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,21 +15,77 @@ type Coin struct {
+// an equal or greater value
 func (c Coin) IsGTE(other Coin) bool {
 	if c.Denom != other.Denom {
-		panic("invalid coin denominations: " + c.Denom)
+		panic("comparing different denominations: " + c.Denom + ", " + other.Denom)
 	}
```
returning errors here would be great for easier unit testing.

leohhhn replied:
Fair point, however:

This was copied over from the tm2 version of coins. The issue with this is that we do not have support for Errorf or fmt in stdlibs, so the dynamic parsing of the errors would be a pain. WDYT?

leohhhn replied:
I suggest we go with this now, and work on a full solution for revamping the errors on both levels in a different effort.

ajnavarro replied:
Good point, but sad. let's leave it as it is.

ajnavarro commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,21 +15,77 @@ type Coin struct {
+	return c.Amount == 0
+}
+
+func NewCoins(val ...interface{}) Coins {
```
Why not just:

```suggestion
func NewCoins(coins ...Coin) Coins {
```

intended to be used like:

```go
NewCoins(NewCoin(20, "DENOM"), NewCoin(1, "DENOM2"))
```

I don't see the benefits of this constructor.

leohhhn replied:
Mainly because the actual calls to this would be 

`std.NewCoins(std.NewCoin(20, "ugnot"), std.NewCoins(20,"xgnot")`

which is more or less the same as, or even more complicated to write than

`std.Coins{std.Coin{20, "ugnot"), std.Coin{20,"xgnot"}}`'

The main benefit is that it is very simple to write. WDYT?

ajnavarro replied:
first of all, defining structs without specifying the types is not recommended, so:

`std.Coins{std.Coin{20, "ugnot"}, std.Coin{20,"xgnot"}}` should be `std.Coins{std.Coin{Amount: 20, Denom: "ugnot"), std.Coin{Amount: 20, Denom: "xgnot"}}`

Now constructors make more sense.

Also, using types in a strongly typed language avoids errors. Using interfaces that need to have specific types in specific positions on a specific number if not will panic, look like a source of problems and bugs.

ajnavarro replied:
We talked about this and we agreed to have two methods: the standard constructor using types, and a specific utility method called `CoinsFromValues(vals ... interface{})` or similar.

ltzmaxwell commented:
> While we're on this topic, I came across this issue I made some time ago: #1676
> 
> Thinking about this a bit, we should really make a good, smooth constructor for the `Coins` struct.
> 
> One approach that I have is the following (in gnovm/stdlibs/std):
> 
> ```
> func NewCoins(val ...interface{}) Coins {
> 	if len(val)%2 != 0 {
> 		panic("invalid number of arguments")
> 	}
> 
> 	cz := Coins{}
> 	for i := 0; i < len(val)-1; i++ {
> 		denom := val[i].(string) // todo validate?
> 		amt := val[i+1].(int64) // todo validate?
> 		cz = append(cz, NewCoin(denom, amt))
> 	}
> 
> 	return cz
> }
> ```
> 
> I'm aware that this is not the safest approach, but it's probably the most comfortable to use. With this, you can just do
> 
> ```
> coins := std.NewCoins("ugnot", 100, "othercoin", 200)
> ```
> 
> and its very painless, instead of `coins:= std.Coins{std.Coin{ "ugnot", coinToSend }})`.
> 
> WDYT?

I like this.

leohhhn commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,21 +15,77 @@ type Coin struct {
 }
 
+// NewCoin returns a new coin with a denomination and amount
+func NewCoin(denom string, amount int64) Coin {
```
This is the ordering in the whole codebase, I believe that if we choose to switch these we should do it in a separate PR.

ajnavarro replied:
I feel it will be a problem soon, but we can leave it as it is for historical reasons...

zivkovicmilos replied:
@ajnavarro +1

zivkovicmilos replied:
@leohhhn 
How are they reversed, there was never a constructor before? We always initialized objects directly (since these are public fields)

I think it's fine to swap them in this PR

leohhhn replied:
As stated in the PR description, this functionality is ported/adapted from `coins.go` in TM2. This is the way they are over there. Sure, I can swap them if needed, but I see no issue with first specifying the denom and then the amount.

zivkovicmilos replied:
Why not swap it for all occurrences?

thehowl replied:
@zivkovicmilos Out of scope

leohhhn commented:
Marking as ready for review for visibility and inputs from the team.

leohhhn commented:
Ran into an issue with regexp: https://github.com/gnolang/gno/issues/2139

leohhhn commented:
@thehowl 

Super weird thing I'm not understanding - since we have std & stdshim and both apparently need to exist right now, how is it possible that at [46a8061](https://github.com/gnolang/gno/pull/2104/commits/46a806189556c75eca85d5319e4b7f7514b6faeb) commit, I've only added changes to the `std` and not to `stdshim`, and the txtar tests are passing and functions working properly? O_O

leohhhn commented:
Ok, the current state is this - we have:

- `Coin.Add`,
- `Coin.Sub`,
- `Coin.IsPositive`,
- `Coin.IsNegative`,
- `Coin.IsZero`

And constructors for `Coin`, as well as two constructors for `Coins` - a "safe" one, and an "unsafe" one ([see discussion](https://github.com/gnolang/gno/pull/2104#discussion_r1604462561))

leohhhn commented:
Will work on updating the docs when I have approvals for these functionalities from the team.

zivkovicmilos commented on gno.land/pkg/integration/testdata/improved-coins.txtar:
```diff
@@ -0,0 +1,94 @@
+gnoland start
+
+gnokey maketx addpkg -pkgdir $WORK -pkgpath gno.land/r/demo/coins -gas-fee 1000000ugnot -gas-wanted 20000000 -broadcast -chainid=tendermint_test test1
```
I think you can use `loadpkg` here

leohhhn replied:
Switched it out: [16109c4](https://github.com/gnolang/gno/pull/2104/commits/16109c4d5cfcbb59f09128bb48fd6de78a0e2465)

zivkovicmilos commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,28 +14,147 @@ type Coin struct {
+
+// String provides a human-readable representation of a coin
 func (c Coin) String() string {
+	if c.IsZero() {
+		return "0" + c.Denom
+	}
```
How is this different than always returning `strconv.Itoa(int(c.Amount)) + c.Denom`?

zivkovicmilos commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,28 +14,147 @@ type Coin struct {
+
+// IsEqual returns true if the two sets of Coins have the same value
+func (c Coin) IsEqual(other Coin) bool {
+	if c.Denom != other.Denom {
+		panic("comparing different denominations: " + c.Denom + ", " + other.Denom)
+	}
```
This check should be extracted as it's very common

leohhhn replied:
[de5f3a3](https://github.com/gnolang/gno/pull/2104/commits/de5f3a39463310df956c79dbb948cbb32400b600)

zivkovicmilos commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,28 +14,147 @@ type Coin struct {
+		panic("coin add overflow/underflow: " + strconv.Itoa(int(c.Amount)) + " + " + strconv.Itoa(int(coinB.Amount)))
+	}
+
+	return Coin{c.Denom, sum}
```
Why does it return a new coin, and not update the existing one?

leohhhn replied:
Changed it: [779ff9a](https://github.com/gnolang/gno/pull/2104/commits/779ff9ad936c4e72960bbbcfe0c74628f850cd0c)

zivkovicmilos commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,28 +14,147 @@ type Coin struct {
+	if !ok {
+		panic("coin add overflow/underflow: " + strconv.Itoa(int(c.Amount)) + " + " + strconv.Itoa(int(coinB.Amount)))
+	}
+	return Coin{c.Denom, dff}
```
Same question as with the addition, I'm not convinced we need to create the new objects every time

leohhhn replied:
Changed it: [779ff9a](https://github.com/gnolang/gno/pull/2104/commits/779ff9ad936c4e72960bbbcfe0c74628f850cd0c)

zivkovicmilos commented on gnovm/stdlibs/stdshim/coins.gno:
```diff
@@ -1,9 +1,12 @@
 
 // NOTE: this is selectly copied over from pkgs/std/coin.go
 // TODO: import all functionality(?).
+// TODO: implement Coin/Coins constructors.
```
Leftover?

leohhhn replied:
Reference comment: https://github.com/gnolang/gno/pull/2104#discussion_r1608432339

zivkovicmilos commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,28 +14,147 @@ type Coin struct {
+
+// NewCoinsFromVal is a type-unsafe way to make a Coins struct
+// Usage is not recommended
+func NewCoinsFromVal(val ...interface{}) Coins {
```
I think we don't need this constructor, it's bloat

leohhhn replied:
[22a171e](https://github.com/gnolang/gno/pull/2104/commits/22a171e5bbc244ee6c2d9b604960cc62862a9601)

zivkovicmilos commented on gnovm/stdlibs/stdshim/coins.gno:
```diff
@@ -12,21 +15,62 @@ type Coin struct {
+
+// String provides a human-readable representation of a coin
 func (c Coin) String() string {
+	if c.IsZero() {
+		return ""
+	}
```
Same comment as to relating `std` code

leohhhn replied:
stdshim was pending a question for @thehowl - now std & stdshim coins.gno are the same: 

[b423d29](https://github.com/gnolang/gno/pull/2104/commits/b423d2944c8600050ce7ad3e8e304caf288ae010)

zivkovicmilos commented on gnovm/stdlibs/stdshim/coins.gno:
```diff
@@ -12,21 +15,62 @@ type Coin struct {
+// IsLT returns true if they are the same type and the receiver is
+// a smaller value
+func (c Coin) IsLT(other Coin) bool {
+	if c.Denom != other.Denom {
+		panic("comparing different denominations: " + c.Denom + ", " + other.Denom)
+	}
```
Same comment as to relating `std` code

leohhhn replied:
Reference comment: https://github.com/gnolang/gno/pull/2104#discussion_r1608432339

zivkovicmilos reviewed (changes requested):
Thank you for adding this functionality 🙏 

There are a few things we need to change before we move forward, please check the comments 🙏 

Additionally, please update the adequate documentation pages as well

zivkovicmilos commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,28 +14,125 @@ type Coin struct {
+// An overflow or underflow panics.
+// An invalid result panics.
+func (c Coin) Sub(coinB Coin) Coin {
+	res := c.SubUnsafe(coinB)
+	return res
```
This can be inlined

leohhhn replied:
Removed the unsafe functions completely, as per @thehowl's comments: [07cf1c9](https://github.com/gnolang/gno/pull/2104/commits/07cf1c9f8589a62a894c12ccf8660ec4b1995f0e)

zivkovicmilos commented on gnovm/stdlibs/stdshim/coins.gno:
```diff
@@ -12,28 +14,125 @@ type Coin struct {
+	return c.Amount == other.Amount
+}
+
+// Adds amounts of two coins with same denom.
```
This comment does not follow standard go comment convention

leohhhn replied:
Added comments: [07cf1c9](https://github.com/gnolang/gno/pull/2104/commits/07cf1c9f8589a62a894c12ccf8660ec4b1995f0e)

zivkovicmilos commented on gnovm/stdlibs/stdshim/coins.gno:
```diff
@@ -12,28 +14,125 @@ type Coin struct {
+	return res
+}
+
+func (c Coin) AddUnsafe(coinB Coin) Coin {
```
Why doesn't this have a comment? Same question for the other methods in the file

leohhhn replied:
Added comments: [07cf1c9](https://github.com/gnolang/gno/pull/2104/commits/07cf1c9f8589a62a894c12ccf8660ec4b1995f0e)

zivkovicmilos commented on gnovm/stdlibs/stdshim/coins.gno:
```diff
@@ -12,28 +14,125 @@ type Coin struct {
+	return c
+}
+
+// Subtracts amounts of two coins with same denom.
```
This comment does not follow standard go comment convention

leohhhn replied:
Added comments: [07cf1c9](https://github.com/gnolang/gno/pull/2104/commits/07cf1c9f8589a62a894c12ccf8660ec4b1995f0e)

zivkovicmilos commented on gnovm/stdlibs/stdshim/coins.gno:
```diff
@@ -43,23 +142,36 @@ func (cz Coins) AmountOf(denom string) int64 {
 }
 
-// TODO implement Coin/Coins constructors.
+// expand for usage within natively bound functions.
```
This comment does not follow standard go comment convention

leohhhn replied:
Added comments: [07cf1c9](https://github.com/gnolang/gno/pull/2104/commits/07cf1c9f8589a62a894c12ccf8660ec4b1995f0e)

zivkovicmilos reviewed (changes requested):
Please see the leftover comments 🙏

thehowl commented:
I did some changes in this PR, starting from wanting to verify that the fix in #2168 did indeed fix the issue we were having here

I saw some build issues and started investigating, but essentially stdshim was using math/overflow, which cannot be used right now because it is not in the stdlibWhitelist. (The error was not printed due to a swallowed error; this is fixed by #1695, but that PR is still stuck in review limbo)

I changed the stdshim version not to use `math/overflow`, so we can safely merge this for now; it matters little anyway (only makes the code "unsafe" for transpiled code, used literally nowhere), and it won't matter entirely after #1695 is merged.

thehowl commented:
There are still some failing tests, but these seem in "userland" so I'll leave you to fix them

thehowl commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,28 +14,125 @@ type Coin struct {
+}
+
+// IsNegative returns true if the coin amount is negative and false otherwise.
+func (c Coin) IsNegative() bool {
```
I have questions on the validity of IsNegative/IsZero, but I know this is just copying the code from the existing implementation, so it's OK for now

thehowl reviewed (commented):
There are some txtar tests failing because they're out of gas, due to the fact that adding `std` now uses more gas; just bump the count where needed.

leohhhn commented on gnovm/stdlibs/std/coins.gno:
```diff
@@ -13,28 +14,125 @@ type Coin struct {
+// If the coins differ in denom then it panics.
+// An overflow or underflow panics.
+// An invalid result panics.
+func (c Coin) Add(coinB Coin) Coin {
```
[07cf1c9](https://github.com/gnolang/gno/pull/2104/commits/07cf1c9f8589a62a894c12ccf8660ec4b1995f0e)

leohhhn commented:
@thehowl thanks for helping out. Fixed the tests, as well as @zivkovicmilos's comments.

leohhhn commented:
Found an issue: https://github.com/gnolang/gno/issues/2204

I implemented a duplicate check on the Coins ctor and added the docs: [c82680a](https://github.com/gnolang/gno/pull/2104/commits/c82680aea455aa74903dc212cad6c9be44a3bbc6)

leohhhn commented:
Not sure why CI is failing.

moul commented on gnovm/stdlibs/stdshim/coins.gno:
```diff
@@ -12,54 +11,164 @@ type Coin struct {
+	return c.Amount < 0
+}
+
+// IsZero returns if this represents no money
```
I would avoid using the term money.

leohhhn replied:
[3762497](https://github.com/gnolang/gno/pull/2104/commits/3762497f04ad6bcaa5fd810b60f356014a261279)

leohhhn commented:
Codecov broken? cc @ajnavarro
